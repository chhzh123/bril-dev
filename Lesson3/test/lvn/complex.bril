@main {
  x: int = const 14;
  y: int = const 2;

  # constant propagation
  div_xy: int = div x y;

  # copy propagation
  div_xy2: int = id div_xy;

  # dead code
  add_xy: int = add x y;

  # common subexpression elimination
  mul_xy: int = mul div_xy div_xy2;

  # reassignment
  y: int = const 8;
  print y;

  # double reassignment
  y: int = const 10;
  print y;

  # further propagation
  sub_xy: int = sub y mul_xy;
  print sub_xy;

  # constant folding
  result: int = const -39;
  equality: bool = eq sub_xy result; 
  print equality;
}