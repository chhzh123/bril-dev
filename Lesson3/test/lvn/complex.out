@main {
  y_new1: int = const 8;
  print y_new1;
  y: int = const 10;
  print y;
  sub_xy: int = const -39;
  print sub_xy;
  equality: bool = const true;
  print equality;
}
